### PR TITLE
feat(arkor): type the createTrainer model field so unsupported base models fail at compile time

### LINKED ISSUE #200: Type the createTrainer model field so unsupported base models fail at compile time
## Summary

The `model` field of `createTrainer` is typed as `string`, but the Arkor backend today accepts **exactly one** value. That mismatch means a typo or an unsupported model name compiles fine and only fails at run time. Narrow the type (and align the runtime schema) so an unsupported model is caught by the TypeScript compiler.

## Current behaviour

- The backend accepts **exactly one** value today: `unsloth/gemma-4-E4B-it`. This is the identifier every starter template (`triage`, `translate`, `redaction`) passes, and the only value Studio's Playground base-model mode accepts. See [`docs/models.mdx`](docs/models.mdx#L6).
- But the types accept any string:
  - [`packages/arkor/src/core/types.ts`](packages/arkor/src/core/types.ts#L306): `TrainerInput.model: string`
  - [`packages/arkor/src/core/types.ts`](packages/arkor/src/core/types.ts#L34): `model: string`
  - [`packages/arkor/src/core/schemas.ts`](packages/arkor/src/core/schemas.ts#L36): `z.looseObject({ model: z.string() })`

So `createTrainer({ model: "unsloth/gema-4", ... })` (typo) type-checks and only blows up when the job runs.

## Proposed behaviour

- Introduce a single source of truth for the supported base models and derive both the TS type and the Zod validator from it, e.g.:

  ```ts
  export const SUPPORTED_MODELS = ["unsloth/gemma-4-E4B-it"] as const;
  export type SupportedModel = (typeof SUPPORTED_MODELS)[number];
  // -> today this is the single literal "unsloth/gemma-4-E4B-it"
  ```

- Type `TrainerInput.model` as `SupportedModel`.
- Replace the loose `z.string()` validation with `z.enum(SUPPORTED_MODELS)` so run time and compile time agree.
- Deriving from one `as const` array means opening the field to the full Gemma 4 family later ([roadmap Backlog](docs/roadmap.mdx#L47)) is a **one-line** change.

## Notes

- The roadmap intentionally plans to widen this list; the derive-from-one-constant shape keeps that cheap. This issue is about making the **current** constraint visible in the type, not about locking the catalog down forever.
- Arkor is in alpha, so narrowing this public field is fine even though it is technically a breaking change for anyone passing a raw string today.

## Deliverables (per AGENTS.md)

- Vitest coverage (a supported model type-checks / validates; an unsupported one is rejected by the schema).
- Docs in **both** languages where the `model` field is described ([`docs/models.mdx`](docs/models.mdx) / [`docs/ja/models.mdx`](docs/ja/models.mdx) and any `createTrainer` API reference pages).

### PR DESCRIPTION
**Pull request by octoarden**
## Summary

Implements #200: `createTrainer`'s `model` is now typed against the list of base models the backend accepts, so a typo or an unsupported model fails at compile time instead of surfacing as a 4xx after the job is submitted.

- New `core/models.ts`: `SUPPORTED_MODELS` (`as const`) plus the derived `SupportedModel` union, both exported from the package root. Mirrors the existing `SUPPORTED_BASE_MODELS` shape in `@arkor/studio-app`'s `lib/baseModels.ts`. The two copies are deliberate: the SPA does not import from `arkor`, so unifying them would pull the SDK into the browser bundle.
- `TrainerInput.model: string` -> `SupportedModel`.
- Widening the catalog later stays a one-line edit to the array, per the issue.

## One deliberate deviation from the issue

The issue also asks to replace `z.string()` with `z.enum(SUPPORTED_MODELS)` in `schemas.ts` and to narrow the second `model: string` in `types.ts` (`JobConfig`). This PR leaves both as `string`, because both sit on the response path, not the input path:

- `trainingJobSchema` backs `jobDetailResponseSchema` and `createJobResponseSchema`, consumed through `decode()` in `client.ts`. It validates jobs coming back from cloud-api, and `decode()` uses `.parse()`, which throws.
- `JobConfig` doubles as `TrainingJob.config`, the decoded job shape (`types.ts`).

Narrowing them to the frozen list would turn any job whose `model` this SDK version predates into a hard `ZodError` at decode time. `trainer.wait()`, `arkor start` (via the runner), and Studio's job routes (which thinly wrap `CloudApiClient`) would all break for users on an older release the moment the backend widens its list, which the roadmap explicitly plans. The backend also normalises model ids internally (see the note in `studio-app/src/lib/baseModels.ts`), so even the echo of a just-created job is not guaranteed to be byte-identical to the input.

The "an unsupported one is rejected by the schema" deliverable is therefore covered at the compiler level instead (`@ts-expect-error` case in `models.test.ts`): there is no input-side runtime validator today, and the response-side one needs to stay tolerant. A regression test in `schemas.test.ts` pins that tolerance so a later pass cannot tighten it by accident.

Narrowing only the input side also keeps the change small: the literal reading of the issue breaks 50 fixture sites across 6 test files, while this shape breaks 24 in one file (all placeholder `model: "m"` inputs in `trainer.test.ts`).

Related call, made the same way: `CloudApiClient.createJob()` still accepts `config: JobConfig` with `model: string`. That is left as-is on purpose. The raw client is the low-level escape hatch, and keeping it permissive means a user on an older SDK can still target a model the backend has already started accepting (the exact window the roadmap's widening creates). `TrainerInput` itself follows the same philosophy with its forwarded-as-is extra fields. The typed guardrail lives on the documented path (`createTrainer`); happy to narrow the raw client too if you prefer consistency over the escape hatch.

## Tests

- `models.test.ts` (new): list shape, a supported model type-checks, an unsupported model is a compile error, and `JobConfig.model` stays assignable from an arbitrary string.
- `schemas.test.ts`: `trainingJobSchema` accepts a `config.model` this SDK version does not know.
- `trainer.test.ts`: input-side fixtures now use a `MODEL` constant derived from `SUPPORTED_MODELS`; response-side fixtures keep arbitrary strings on purpose.

## Docs

EN/JA pairs updated in the same PR per AGENTS.md: `models.mdx`, `sdk/create-trainer.mdx`, and `cookbook/customizing-templates.mdx` (its "forwarded to the cloud API as a string / 4xx from upstream" paragraph described the pre-change behaviour and was rewritten).

## Breaking change

`TrainerInput.model` narrows a public field. Alpha per CONTRIBUTING; the issue explicitly opts into this.

## Verification

All gates green locally on macOS, Node 24.18.1, pnpm 11.13.1: `check:no-em-dash`, `format:check`, `typecheck` (10/10 tasks), `lint` (0 warnings, 0 errors across 7 packages), unit suites (arkor 451, cli-internal 224, studio-app 205, create-arkor 47), e2e-cli (110 passed), e2e-studio (9 passed).

Noting there was an assignment request on #200 without follow-up; happy to hand this off if that work is already in flight.

---
## Summary by cubic
Types `createTrainer`’s `model` as `SupportedModel` and enforces a frozen `SUPPORTED_MODELS` list at construction, so unsupported IDs fail fast in both TS and JS/CLI. Response shapes (`JobConfig.model` and schemas) stay `string` for forward compatibility. Implements #200.

- **New Features**
  - Added frozen `SUPPORTED_MODELS` and `SupportedModel` in `core/models.ts`, exported from `arkor`.
  - Narrowed `TrainerInput.model` to `SupportedModel` and added a runtime guard in `createTrainer`; the raw client stays permissive (`CloudApiClient.createJob` with `JobConfig.model: string`).

- **Migration**
  - Update trainer code to use a supported ID like `"unsloth/gemma-4-E4B-it"` or select from `SUPPORTED_MODELS`.

<sup>Written for commit ec6fb788b2ed4aab7d6a3e95b31adbde3f4b304c. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/arkorlab/arkor/pull/211?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by CodeRabbit

* **New Features**
  * Added a supported-model list and `SupportedModel` type to the SDK.
  * Trainer creation now validates model selection at compile time and rejects unsupported models immediately at runtime.
  * Added public exports for supported model values and types.
  * Cloud job configurations continue to support server-provided model identifiers.
* **Documentation**
  * Updated SDK and model guides with supported-model usage, validation behavior, and examples.
* **Tests**
  * Added coverage for model validation and compatibility with server-provided identifiers.

### CODE REVIEW

**Comment by octoarden** [comment]
Pushed 1883f18 addressing the review:

- **Codex / cubic P2** (the esbuild path never typechecks): `createTrainer` now validates `model` against `SUPPORTED_MODELS` at construction time and throws with a clear message, so `arkor build` / `arkor start` and plain-JavaScript callers fail fast locally instead of hitting a backend 4xx. The raw `CloudApiClient.createJob` escape hatch stays permissive as described in the PR body.
- **CodeRabbit**: the compile-time tests now assert `Parameters<typeof createTrainer>[0]["model"]` (the public boundary), and the docs keep the full diagnostic on the `model` line (EN/JA).
- **cubic P3**: the id-shape guard now requires the `owner/model` form.

On the suggested parity test between `SUPPORTED_MODELS` and `@arkor/studio-app`'s `SUPPORTED_BASE_MODELS`: declining for now. `arkor`'s build already consumes `studio-app`'s dist (copy-studio-assets), so a reverse devDependency would create a workspace cycle, and a relative-path import across package roots would cross the per-package tsconfig boundaries. Both lists carry mirrored comments pointing at each other; happy to wire a generated shared artifact in a follow-up if you'd prefer that.

**Comment by octoarden** [comment]
@coderabbitai review

(f3d3ef8 also reconciles the create-trainer intro with the new runtime guard, per cubic's follow-up.)

**Comment by k-taro56** [comment]
@coderabbitai review

**Comment by octoarden** [comment]
The one red job (windows-latest, node >=24.0.0 <24.1.0) is a 5s timeout in `server.test.ts`'s `/api/manifest` test, which this PR does not touch (its fixture is a hand-rolled manifest object; `createTrainer` is never called in that file). Looks like the strip-types RC row being slow on the dynamic TS import; the same job passes on every other row. A rerun should clear it whenever convenient.

File: docs/sdk/create-trainer.mdx [hunk 1]
```diff
@@ -17,7 +17,7 @@ export const trainer = createTrainer({
 });
 ```
 
-`createTrainer` returns a `Trainer` object with `start` / `wait` / `cancel` methods (see [Trainer control](/sdk/trainer-control)). The function does **not** run client-side validation on `TrainerInput`: TypeScript checks the typed fields at compile time, and the rest is forwarded to the cloud-api when `start()` runs. Bad values surface as backend errors during the run, not as client-side throws at construction.
+`createTrainer` returns a `Trainer` object with `start` / `wait` / `cancel` methods (see [Trainer control](/sdk/trainer-control)). With one exception, the function does **not** run client-side validation on `TrainerInput`: TypeScript checks the typed fields at compile time, and the rest is forwarded to the cloud-api when `start()` submits the job, so bad values surface as backend errors at job creation or during the run. The exception is `model`, which is checked against [`SUPPORTED_MODELS`](/models) at construction and throws immediately on an unsupported value (the `arkor build` / `arkor start` path bundles with esbuild and never typechecks, so the compile-time narrowing alone cannot gate it).
```
[the remark this answers is not in the document]
Comment (octoarden) [reply docs/sdk/create-trainer.mdx]: The current wording is deliberate. Job creation only shape-checks the config: the server-side schema for it is a looseObject (see the comment inside `CloudApiClient.createJob` in `packages/arkor/src/core/client.ts`: any object with a `model` and a `datasetSource` passes). A value like a mistyped dataset name therefore clears job creation and only fails once the worker consumes it, surfacing as a `training.failed` event during the run. "At job creation or during the run" covers both real timings; narrowing it to job creation would make the page claim a validation the backend does not perform. Leaving as is.

**Comment by octoarden** [comment]
@k-taro56 **TL;DR:** I pinged you because CodeRabbit keeps asking me to change wording this PR barely touched, and its latest request would make the docs wrong. I cannot talk to the bot as a non-org member, so please either dismiss its review or tell me to follow it, and I will.

---

Full state summary, since the bot threads got long:

- All review-bot findings are addressed (4 commits) except two I declined with reasons in-thread:
  1. the `SUPPORTED_MODELS` / studio-app parity test: wiring it would need a studio-app -> arkor dev dependency, which is a workspace cycle since `arkor`'s build already copies studio-app's dist.
  2. CodeRabbit's latest wording request on `create-trainer.mdx`: taking it would make the page claim job-creation-time validation the backend does not perform (the job-config schema is a looseObject per the comment in `CloudApiClient.createJob`, so some bad values only surface as `training.failed` during the run).
- CodeRabbit's current Changes Requested is only about point 2, and as a non-org member I cannot interact with it, so it needs your dismissal either way. If you would rather I just take its wording, say so and I will.
- The one red CI job (windows, node >=24.0.0 <24.1.0) is a 5s timeout in `server.test.ts`'s `/api/manifest` test, which this PR does not touch; analysis in my earlier comment.
- The latest push (d17c403, docs only) is waiting for workflow approval.

Happy to change course on any of these.

**Comment by octoarden** [comment]
I'll just try making CodeRabbit happy a little bit more.

File: docs/ja/sdk/create-trainer.mdx [hunk 2]
```diff
@@ -17,7 +17,7 @@ export const trainer = createTrainer({
 });
 ```
 
-`createTrainer` は `start` / `wait` / `cancel` メソッドを持つ `Trainer` オブジェクトを返します（[トレーナー制御](/ja/sdk/trainer-control) を参照）。この関数は `TrainerInput` のクライアント側バリデーションを **しません**: TypeScript が型付きフィールドをコンパイル時にチェックし、それ以外は `start()` 実行時にクラウド API へ転送されます。不正な値は構築時のクライアント側 throw ではなく、学習中にバックエンドエラーとして現れます。
+`createTrainer` は `start` / `wait` / `cancel` メソッドを持つ `Trainer` オブジェクトを返します（[トレーナー制御](/ja/sdk/trainer-control) を参照）。1 つの例外を除き、この関数は `TrainerInput` のクライアント側バリデーションを **しません**: TypeScript が型付きフィールドをコンパイル時にチェックし、それ以外は `start()` がジョブを作成するときにクラウド API へ転送されるため、不正な値はジョブ作成の時点からバックエンドエラーとして現れます。例外は `model` で、生成時に [`SUPPORTED_MODELS`](/ja/models) と照合され、サポートされない値では即座に例外を投げます（`arkor build` / `arkor start` の経路は esbuild でバンドルされ型チェックが走らないため、コンパイル時の絞り込みだけではゲートできません）。
```
[the remark this answers is not in the document]
Comment (octoarden) [reply docs/ja/sdk/create-trainer.mdx]: This is the same timing question CodeRabbit raised in the opposite direction: its review on this line asked me to drop the previous 「ジョブ作成時または実行中に」 wording, which this comment now endorses. The current JA 「ジョブ作成の時点から現れます」 is intended to carry the same onward semantics as the EN "once the job is created": from creation on, including failures that only appear later as `training.failed`. Since the two bot reviews now contradict each other on this sentence, I am leaving the wording as is unless a maintainer prefers otherwise.

**Comment by soleil-colza** [comment]
@coderabbitai resolve

**Review by soleil-colza — APPROVED** [review]
Thanks for your contribution!!
[LGTMeow](https://lgtmeow.com)

File: docs/ja/models.mdx [hunk 3]
```diff
@@ -19,6 +19,29 @@ export const trainer = createTrainer({
 });
 ```
 
+## リストはコンパイラが強制します
```
Comment (k-taro56) [inline docs/ja/models.mdx]:
```suggestion
## リストはコンパイラーが強制します
```

File: packages/arkor/src/core/types.ts [hunk 4]
```diff
@@ -302,8 +309,8 @@ export interface LoraConfig {
 export interface TrainerInput {
   /** Human-readable run name; shown in Studio + Web UI. */
   name: string;
-  /** Base model identifier (HuggingFace path, etc.). */
-  model: string;
+  /** Base model identifier. Narrowed to what cloud-api accepts today. */
+  model: SupportedModel;
```
[the remark this answers is not in the document]
Comment (octoarden) [reply packages/arkor/src/core/types.ts:313]: Addressed in ec6fb78.

File: packages/arkor/src/core/models.ts [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+ * Widening the array is the only edit needed to open the field up to the rest
+ * of the Gemma 4 family (see the roadmap Backlog): the `SupportedModel` union
+ * is derived from it.
+ *
+ * Studio's Playground keeps its own copy of this list in
+ * `@arkor/studio-app`'s `lib/baseModels.ts`. That duplication is deliberate:
+ * the SPA bundle does not import from `arkor`, so unifying the two would pull
+ * the SDK into the browser bundle. Keep them in sync by hand.
+ */
+export const SUPPORTED_MODELS = ["unsloth/gemma-4-E4B-it"] as const;
```
[the remark this answers is not in the document]
Comment (octoarden) [reply packages/arkor/src/core/models.ts]: Addressed in ec6fb78.
